lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,913 @@
+			select {
+			case errChan <- err:
+			case <-ctxt.Done():
+			}
+
+			return
+		}
+
+		select {
+		case registerChan <- msg:
+		case <-ctxt.Done():
+		}
+	}()
+
+	// Wait for the initial message to arrive or time out if it takes too
+	// long. The initial message must be a registration message from the
+	// remote signer.
+	var registrationMsg *signerResponse
+	select {
+	case registrationMsg = <-registerChan:
+		if registrationMsg.GetRefRequestId() != handshakeRequestID {
+			return fmt.Errorf("initial request ID must be %d, "+
+				"but is: %d", handshakeRequestID,
+				registrationMsg.GetRefRequestId())
+		}
+
+		register := registrationMsg.GetSignerRegistration()
+
+		// TODO(viktor): This could be extended to validate the version
+		// of the remote signer in the future.
+		if register == "" {
+			return errors.New("invalid initial remote signer " +
+				"registration message")
+		}
+
+	case err := <-errChan:
+		return fmt.Errorf("error receiving initial remote signer "+
+			"registration message: %v", err)
+
+	case <-s.quit:
+		return ErrShuttingDown
+
+	case <-ctxt.Done():
+		return ctxt.Err()
+	}
+
+	// Send a message to the client to indicate that the registration has
+	// successfully completed.
+	req := &walletrpc.SignCoordinatorRequest_RegistrationComplete{
+		RegistrationComplete: true,
+	}
+
+	regCompleteMsg := &walletrpc.SignCoordinatorRequest{
+		RequestId:       handshakeRequestID,
+		SignRequestType: req,
+	}
+
+	// Send the message in a goroutine because the Send method blocks until
+	// the message is read by the client.
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+
+		err := stream.Send(regCompleteMsg)
+		if err != nil {
+			select {
+			case errChan <- err:
+			case <-ctxt.Done():
+			}
+
+			return
+		}
+
+		close(registerDoneChan)
+	}()
+
+	select {
+	case err := <-errChan:
+		return fmt.Errorf("error sending registration complete "+
+			" message to remote signer: %v", err)
+
+	case <-ctxt.Done():
+		return ctxt.Err()
+
+	case <-s.quit:
+		return ErrShuttingDown
+
+	case <-registerDoneChan:
+	}
+
+	return nil
+}
+
+// StartReceiving is the main receive loop that receives responses from the
+// remote signer. Responses must have a RequestID that corresponds to requests
+// which are waiting for a response; otherwise, the response is ignored.
+func (s *SignCoordinator) StartReceiving() {
+	defer s.wg.Done()
+
+	// Signals to any ongoing requests that the remote signer is no longer
+	// connected.
+	defer close(s.doneReceiving)
+
+	for {
+		resp, err := s.stream.Recv()
+		if err != nil {
+			select {
+			case <-s.quit:
+				// If we've already shut down, the main Run
+				// method will not be able to receive any error
+				// sent over the error channel. So we just
+				// return.
+
+			case s.receiveErrChan <- err:
+				// Send the error over the error channel, so
+				// that the main Run method can return the error
+			}
+
+			return
+		}
+
+		s.mu.Lock()
+
+		respChan, ok := s.responses[resp.GetRefRequestId()]
```

Comment:
Awesome suggestion, that simplified this quite a bit! I've updated this to use a `SyncMap` instead. The main mutex for `SignCoordinator` needs to remain for now though, to avoid that we get a concurrent connection to watch-only node at the exact time that we `Stop` the `SignCoordinator` (which could potentially execute `s.wg.Add(1)` in `Run` after the `s.wg.Wait()` has just stopped blocking in `Stop`).